TombEngine/Objects/TR3/Trap/ElectricCleaner.cpp

```diff
@@ -0,0 +1,380 @@
+			//Is a 2 click step (higher or lower).
+			int distanceToFloor = abs(col.Position.Floor - item.Pose.Position.y);
+			if (distanceToFloor > CLICK(2))
+				return false;
+
+			short slopeAngle = ANGLE(0.0f);
+
+			if (col.FloorTilt.x > 0)
+				slopeAngle = -ANGLE(90.0f);
+			else if (col.FloorTilt.x < 0)
+				slopeAngle = ANGLE(90.0f);
+
+			if (col.FloorTilt.y > 0 && col.FloorTilt.y > abs(col.FloorTilt.x))
+				slopeAngle = ANGLE(180.0f);
+			else if (col.FloorTilt.y < 0 && -col.FloorTilt.y > abs(col.FloorTilt.x))
+				slopeAngle = ANGLE(0.0f);
+
+			auto angleDir = FROM_RAD(atan2(Dir.x, Dir.z));
+			auto alignment = slopeAngle - angleDir;
+
+			//Is slope not aligned with the direction?
+			if ((alignment != 32768) && (alignment != 0) && (alignment != -32768))
+				return false;
+		}
+
+		//Is diagonal floor?
+		if (col.Position.DiagonalStep)
+			return false;
+
+		//Is ceiling (square or diagonal) high enough?
+		int distanceToCeiling = abs(col.Position.Ceiling - col.Position.Floor);	
+		int cleanerHeight = 1024; //TODO change it for the collision bounding box height.
+		if (distanceToCeiling < cleanerHeight)
+			return false;
+
+		//Is a non walkable tile? (So there is not any box)
+		if (col.Block->Box == NO_BOX)
+			return false;
+
+		//Is a blocked grey box (So it's an Isolated box)
+		if (g_Level.Boxes[col.Block->Box].flags & BLOCKABLE)
+			return false;
+		
+		//Is a stopper tile? (There is still a shatter object).
+		if (col.Block->Stopper)
+			return false;
+
+		//If nothing of that happened, then it must be a valid sector.
+		return true;
+	}
+
+	Vector3 ElectricCleanerSearchDirections(ItemInfo& item, const Vector3& Dir1, const Vector3& Dir2, const Vector3& Dir3)
+	{
+		if (IsNextSectorValid(item, Dir1))
+			return Dir1;
+		if (IsNextSectorValid(item, Dir2))
+			return Dir2;
+		if (IsNextSectorValid(item, Dir3))
+			return Dir3;
+
+		return Vector3::Zero;
+	}
+
+	void ElectricCleanerCollision(short itemNumber, ItemInfo* laraItem, CollisionInfo* coll)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		ObjectCollision(itemNumber, laraItem, coll);
+
+		if (item.TouchBits.Test(ElectricCleanerHarmJoints) && item.ItemFlags[2])
+		{
+			ItemElectricBurn(laraItem, -1);
+			laraItem->HitPoints = 0;
+
+			bool flagStopAfterKill = ((item.ItemFlags[1] & (1 << 4)) != 0);
+			if (flagStopAfterKill)
+				item.ItemFlags[2] = 0;
+
+			SoundEffect(SFX_TR3_CLEANER_FUSEBOX, &item.Pose);
+		}
+	}
+		
+	void ElectricCleanerToItemCollision(ItemInfo& item)
+	{
+		auto backupPos = item.Pose.Position;
+
+		switch (item.Pose.Orientation.y)
+		{
+		case ANGLE(0.0f):
+			item.Pose.Position.z += BLOCK(0.5f);
+			break;
+
+		case ANGLE(90.0f):
+			item.Pose.Position.x += BLOCK(0.5f);
+			break;
+
+		case ANGLE(-180.0f):
+			item.Pose.Position.z -= BLOCK(0.5f);
+			break;
+
+		case ANGLE(-90.0f):
+			item.Pose.Position.x -= BLOCK(0.5f);
+			break;
+		}
+
+		if (GetCollidedObjects(&item, CLICK(1), true, CollidedItems, CollidedMeshes, true))
+		{
+			int lp = 0;
```

Comment:
what is lp? :D

Reply:
Good question. I haven't got any idea. XD

That's a function I did not touch, however looks like this **lp** is just acting as an index, (same as if I call it **i** or **currentItem** in the logic I saw nothing wrong, although I don't know why the choice of word **lp**